cirq-core/cirq/protocols/kraus_protocol.py

```diff
@@ -190,28 +207,71 @@ def has_kraus(val: Any, *, allow_decompose: bool = True) -> bool:
             the result is skipped.
 
     Returns:
-        If `val` has a `_has_kraus_` method and its result is not
-        NotImplemented, that result is returned. Otherwise, if `val` has a
-        `_has_mixture_` method and its result is not NotImplemented, that
-        result is returned. Otherwise if `val` has a `_has_unitary_` method
-        and its results is not NotImplemented, that result is returned.
-        Otherwise, if the value has a _kraus_ method return if that
-        has a non-default value. Returns False if none of these functions
-        exists.
+        Whether or not `val` has a Kraus representation.
     """
-    kraus_getter = getattr(val, '_has_kraus_', None)
-    result = NotImplemented if kraus_getter is None else kraus_getter()
-    if result is not NotImplemented:
-        return result
+    result = _gettr_helper(val, ['_has_kraus_', '_has_channel_', '_kraus_'])
+    if result is not None and result is not NotImplemented and result:
+        return True
 
-    result = has_mixture(val, allow_decompose=False)
-    if result is not NotImplemented and result:
-        return result
+    if mixture_protocol.has_mixture(val, allow_decompose=False):
+        return True
 
     if allow_decompose:
         operations, _, _ = _try_decompose_into_operations_and_qubits(val)
         if operations is not None:
             return all(has_kraus(val) for val in operations)
 
-    # No has methods, use `_kraus_` or delegates instead.
-    return kraus(val, None) is not None
+    # No has methods
+    return False
+
+
+def _check_equality(x, y):
+    if type(x) != type(y):
+        return False
+    if type(x) not in [list, tuple, np.ndarray]:
+        return x == y
+    if type(x) == np.ndarray:
+        return x.shape == y.shape and np.all(x == y)
+    return False if len(x) != len(y) else all([_check_equality(a, b) for a, b in zip(x, y)])
+
+
+def _kraus_to_superoperator(op, qubits, default):
+    kraus_list = kraus(op, default)
+    if _check_equality(kraus_list, default):
+        return default
+
+    val = None
+    op_q = op.qubits
+    found = False
+    for i in range(len(qubits)):
+        if qubits[i] in op_q:
+            if not found:
+                found = True
+                if val is None:
+                    val = kraus_list
+                else:
+                    val = tuple([np.kron(x, y) for x in val for y in kraus_list])
+
+        elif val is None:
+            val = (np.identity(2),)
+        else:
+            val = tuple([np.kron(x, np.identity(2)) for x in val])
```

Comment:
I think this would not work for cases where qubits in `op_q` do not occur consecutively in the `qubits` array. For example, if qubits array is [1, 2, 3] and op is a CNOT on (1, 3) ? IIUC, it would behave as if op is a CNOT on (1, 2) instead ?